synapse/handlers/sync.py

```diff
@@ -683,12 +683,11 @@ async def get_state_at(
             stream_position: point at which to get state
             state_filter: The state filter used to fetch state from the database.
         """
-        # FIXME this claims to get the state at a stream position, but
-        # get_recent_events_for_room operates by topo ordering. This therefore
-        # does not reliably give you the state at the given stream position.
-        # (https://github.com/matrix-org/synapse/issues/3305)
         last_events, _ = await self.store.get_recent_events_for_room(
-            room_id, end_token=stream_position.room_key, limit=1
+            room_id,
+            end_token=stream_position.room_key,
+            limit=1,
+            order_by="stream",
         )
```

Comment:
We should add a test for https://github.com/matrix-org/synapse/issues/12281 so it doesn't regress.

[Other MSC2716 `/batch_send` test](https://github.com/matrix-org/synapse/blob/287a9c1e20e1e271d69be581d6bb8ee5a8de9d88/tests/rest/client/test_room_batch.py) for reference. We can also use the barebones reproduction steps explained in https://github.com/matrix-org/synapse/issues/12281#issuecomment-1081199403 for the test.

I think it would be possible write a [Complement](https://github.com/matrix-org/complement/pull/352) test instead but since the problem is so closely tied to `stream_ordering`, it would probably be best to lookup and manipulate from a trial unit test.

Reply:
Pulled together [a test case here](https://github.com/matrix-org/synapse/pull/12319/commits/e440db94d638b778a5a6b7ac2dc909cb37093445) - not sure this is the right location for it as it could sit in sync or batch.